File: server/app/services/scoring/domain_calculator.rb
# frozen_string_literal: true

module Scoring
  # Calculates raw scores per domain from an assessment's responses.
  #
  # Each domain (energy, decision_making, relationship, recovery) has a set of
  # questions scored on a 1-5 Likert scale. Questions with negative polarity
  # are reverse-scored (6 - value) so that higher raw scores consistently
  # indicate one end of the spectrum across all items.
  #
  # Skipped responses (value: nil) are excluded from the sum. The returned
  # hash maps each domain to its raw score total.
  #
  # Usage:
  #   result = Scoring::DomainCalculator.new(assessment).call
  #   # => { "energy" => 18, "decision_making" => 12, ... }
  #
  class DomainCalculator
    DOMAINS = Question::DOMAINS

    attr_reader :assessment

    def initialize(assessment)
      @assessment = assessment
    end

    def call
      DOMAINS.each_with_object({}) do |domain, scores|
        scores[domain] = raw_score_for(domain)
      end
    end

    private

    def raw_score_for(domain)
      domain_responses(domain).sum do |response|
        effective_value(response)
      end
    end

    def domain_responses(domain)
      responses_with_questions.select do |response|
        response.question.domain == domain && !response.skipped?
      end
    end

    def effective_value(response)
      value = response.value
      return 0 if value.nil?

      response.question.negative? ? (6 - value) : value
    end

    # Eager-load questions once to avoid N+1 queries across all domains.
    def responses_with_questions
      @responses_with_questions ||= assessment.responses.includes(:question).to_a
    end
  end
end

File: apps/workers/src/services/scoring/domainCalculator.ts
/**
 * src/services/scoring/domainCalculator.ts — GREEN phase
 * Rails: server/app/services/scoring/domain_calculator.rb
 *
 * calculateDomainScores(assessment) → { energy, decision_making, relationship, recovery }
 * - positive polarity: sum values directly
 * - negative polarity: reverse-score (6 - value)
 * - skipped (null/undefined value): excluded from sum
 * - empty assessment: 0 for all domains
 */

export interface DomainScores {
  energy: number;
  decision_making: number;
  relationship: number;
  recovery: number;
}

export interface ResponseRow {
  value: number | null;
  polarity: string;
  domain: string;
}

export interface AssessmentWithResponses {
  id: number;
  responses: ResponseRow[];
}

type DomainKey = keyof DomainScores;
const DOMAINS: DomainKey[] = ["energy", "decision_making", "relationship", "recovery"];

export function calculateDomainScores(
  assessment: AssessmentWithResponses
): DomainScores {
  const result: DomainScores = {
    energy: 0,
    decision_making: 0,
    relationship: 0,
    recovery: 0,
  };

  for (const response of assessment.responses) {
    const domain = response.domain as DomainKey;
    if (!DOMAINS.includes(domain)) continue;
    if (response.value == null) continue;

    const value =
      response.polarity === "negative"
        ? 6 - response.value
        : response.value;

    result[domain] += value;
  }

  return result;
}
